Auth0Tests/Generators.swift

```diff
@@ -0,0 +1,198 @@
+                             authTime: Date?) -> String {
+    var bodyDict: [String: Any] = [:]
+    
+    if let iss = iss {
+        bodyDict["iss"] = iss
+    }
+    
+    if let sub = sub {
+        bodyDict["sub"] = sub
+    }
+    
+    if let aud = aud {
+        bodyDict["aud"] = aud.count == 1 ? aud[0] : aud
+    }
+    
+    if let exp = exp {
+        bodyDict["exp"] = exp.timeIntervalSince1970
+    }
+    
+    if let iat = iat {
+        bodyDict["iat"] = iat.timeIntervalSince1970
+    }
+    
+    if let azp = azp {
+        bodyDict["azp"] = azp
+    }
+    
+    if let maxAge = maxAge {
+        bodyDict["max_age"] = maxAge
+    }
+    
+    if let authTime = authTime {
+        bodyDict["auth_time"] = authTime.timeIntervalSince1970
+    }
+    
+    if let nonce = nonce {
+        bodyDict["nonce"] = nonce
+    }
+    
+    return encodeJWTPart(from: bodyDict)
+}
+
+func generateJWT(alg: String = JWTAlgorithm.rs256.rawValue,
+                 kid: String = "key123",
+                 iss: String? = "https://tokens-test.auth0.com/",
+                 sub: String? = "auth0|123456789",
+                 aud: [String]? = ["tokens-test-123"],
+                 exp: Date? = Date().addingTimeInterval(100000),
+                 iat: Date? = Date().addingTimeInterval(-1000),
+                 azp: String? = nil,
+                 nonce: String? = "a1b2c3d4e5",
+                 maxAge: Int? = nil,
+                 authTime: Date? = nil,
+                 signature: String? = nil) -> JWT {
+    let header = generateJWTHeader(alg: alg, kid: kid)
+    let body = generateJWTBody(alg: alg,
+                               kid: kid,
+                               iss: iss,
+                               sub: sub,
+                               aud: aud,
+                               exp: exp,
+                               iat: iat,
+                               azp: azp,
+                               nonce: nonce,
+                               maxAge: maxAge,
+                               authTime: authTime)
+    
+    let signableParts = "\(header).\(body)"
+    var signaturePart = "SIGNATURE"
+    
+    if let signature = signature {
+        signaturePart = signature
+    }
+    else if let algorithm = JWTAlgorithm(rawValue: alg) {
+        signaturePart = algorithm.sign(value: signableParts.data(using: .utf8)!).a0_encodeBase64URLSafe()!
+    }
+    
+    return try! decode(jwt: "\(signableParts).\(signaturePart)")
+}
+
+// MARK: - JWK
+
+private func calculateLength(from bytes: UnsafePointer<UInt8>) -> (UnsafePointer<UInt8>, Int) {
+    guard bytes.pointee > 0x7f else { return (bytes + 1, Int(bytes.pointee)) }
+
+    let count = Int(bytes.pointee & 0x7f)
+    let length = (1...count).reduce(0) { ($0 << 8) + Int(bytes[$1]) }
+    let pointer = bytes + (1 + count)
+    
+    return (pointer, length)
+}
+
+private func extractData(from bytes: UnsafePointer<UInt8>) -> (UnsafePointer<UInt8>, Data)? {
```

Comment:
This method extracts the data portion from a given DER triplet. It starts at the triplet's base address, then proceeds from there. It returns a pointer to the end of the triplet plus its data.

Reply:
Comment added in 123af59498094926539a355378c5ce78efa7a1bb.